File: examples/typescript/complex/deep_flow_helpers.ts
/**
 * Deep cross-file flow chain: Middleware/helper classes.
 *
 * DfcValidator checks command structure (weak, no sanitization).
 * DfcTransformPipeline applies transformations preserving taint.
 * DfcCommandRegistry maps verbs to executable paths.
 */

interface DfcParsedCommand {
  verb: string;
  target: string;
  options: string[];
  rawLine: string;
}

interface DfcRegistryEntry {
  verb: string;
  executablePath: string;
  fullCommand: string;
}

export class DfcValidator {
  private allowedVerbs: string[];
  private log: string[];

  constructor() {
    this.allowedVerbs = ["run", "exec", "deploy", "build", "test"];
    this.log = [];
  }

  checkVerb(cmd: DfcParsedCommand): DfcParsedCommand {
    // Step 13: Check verb against allowed list (passes through regardless)
    const found = this.allowedVerbs.indexOf(cmd.verb) >= 0;
    this.log.push("verb-check:" + cmd.verb + ":" + String(found));
    // Bug: returns cmd even if verb not found
    return cmd;
  }

  checkTarget(cmd: DfcParsedCommand): DfcParsedCommand {
    // Step 14: Check target is non-empty (no sanitization)
    const target = cmd.target;
    if (target.length === 0) {
      this.log.push("target-check:empty");
    }
    // Step 15: Shallow copy preserves tainted data
    const validated: DfcParsedCommand = {
      verb: cmd.verb,
      target: target,
      options: cmd.options.slice(),
      rawLine: cmd.rawLine,
    };
    // Step 16: Log validation result
    this.log.push("target-check:" + target + ":ok");
    return validated;
  }
}

export class DfcTransformPipeline {
  private stages: Array<(cmd: DfcParsedCommand) => DfcParsedCommand>;

  constructor() {
    this.stages = [];
    this.stages.push(this._expandPaths.bind(this));
    this.stages.push(this._joinOptions.bind(this));
    this.stages.push(this._buildCommandLine.bind(this));
  }

  private _expandPaths(cmd: DfcParsedCommand): DfcParsedCommand {
    // Step 18: Expand tilde in target (tainted data preserved)
    const expanded = cmd.target.replace("~", "/home/user");
    return {
      verb: cmd.verb,
      target: expanded,
      options: cmd.options,
      rawLine: cmd.rawLine,
    };
  }

  private _joinOptions(cmd: DfcParsedCommand): DfcParsedCommand {
    // Step 19: Join options into target string
    const joined = cmd.options.join(" ");
    // Step 20: Append joined options to target
    const combined = cmd.target + " " + joined;
    return {
      verb: cmd.verb,
      target: combined.trim(),
      options: cmd.options,
      rawLine: cmd.rawLine,
    };
  }

  private _buildCommandLine(cmd: DfcParsedCommand): DfcParsedCommand {
    // Step 21: Build full command line from verb and target
    const fullLine = cmd.verb + " " + cmd.target;
    // Step 22: Store in rawLine for downstream use
    return {
      verb: cmd.verb,
      target: cmd.target,
      options: cmd.options,
      rawLine: fullLine,
    };
  }

  process(cmd: DfcParsedCommand): DfcParsedCommand {
    // Steps 18-22: Run all transform stages
    let current = cmd;
    for (const stage of this.stages) {
      current = stage(current);
    }
    return current;
  }
}

export class DfcCommandRegistry {
  private entries: Map<string, DfcRegistryEntry>;

  constructor() {
    this.entries = new Map();
  }

  register(cmd: DfcParsedCommand): void {
    // Step 24: Build registry entry from parsed command
    const execPath = "/usr/bin/" + cmd.verb;
    // Step 25: Build full command string (tainted target embedded)
    const fullCmd = execPath + " " + cmd.target;
    const entry: DfcRegistryEntry = {
      verb: cmd.verb,
      executablePath: execPath,
      fullCommand: fullCmd,
    };
    // Step 26: Store in registry
    this.entries.set(cmd.verb, entry);
  }

  resolve(verb: string): string {
    // Step 27: Look up entry by verb
    const entry = this.entries.get(verb);
    if (!entry) {
      return verb;
    }
    // Step 28: Return the full command string (tainted)
    return entry.fullCommand;
  }
}

File: examples/c/complex/deep_flow_helpers.c
/* deep_flow_helpers.c -- Middleware/helper functions.
 *
 * DfcValidator checks command structure (weak, no sanitization).
 * DfcTransformPipeline applies transformations preserving taint.
 * DfcCommandRegistry maps verbs to executable paths.
 *
 * All types use "Dfc" prefix. Standard library only.
 */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "deep_flow_chain.h"

/* --- Validation (Steps 12-16) --- */

/* Step 13: Check verb against allowed list */
int dfc_check_verb(struct DfcParsedCommand *cmd) {
    const char *allowed[] = {"run", "exec", "deploy", "build", "test", NULL};
    for (int i = 0; allowed[i] != NULL; i++) {
        if (strcmp(cmd->verb, allowed[i]) == 0) {
            return 1;
        }
    }
    /* Bug: returns 1 anyway -- allows everything through */
    return 1;
}

/* Step 14: Check target is non-empty (no sanitization) */
int dfc_check_target(struct DfcParsedCommand *cmd) {
    /* Step 15: No real sanitization, just length check */
    if (strlen(cmd->target) == 0) {
        return 0;
    }
    /* Step 16: Target passes through tainted */
    return 1;
}

/* Step 12: Validate the parsed command */
int dfc_validate(struct DfcParsedCommand *cmd) {
    if (!dfc_check_verb(cmd)) {
        return 0;
    }
    if (!dfc_check_target(cmd)) {
        return 0;
    }
    return 1;
}

/* --- Transform (Steps 17-22) --- */

/* Step 18: Expand tilde in target (tainted data preserved) */
void dfc_expand_paths(struct DfcParsedCommand *cmd) {
    char *tilde = strchr(cmd->target, '~');
    if (tilde != NULL) {
        char expanded[256];
        size_t prefix_len = (size_t)(tilde - cmd->target);
        snprintf(expanded, sizeof(expanded), "%.*s/home/user%s",
                 (int)prefix_len, cmd->target, tilde + 1);
        strncpy(cmd->target, expanded, sizeof(cmd->target) - 1);
        cmd->target[sizeof(cmd->target) - 1] = '\0';
    }
}

/* Step 19-20: Join options into target string */
void dfc_join_options(struct DfcParsedCommand *cmd) {
    /* Step 19: Read current target */
    char combined[512];
    /* Step 20: Append options to target */
    snprintf(combined, sizeof(combined), "%s %s", cmd->target, cmd->options);
    strncpy(cmd->target, combined, sizeof(cmd->target) - 1);
    cmd->target[sizeof(cmd->target) - 1] = '\0';
}

/* Step 21-22: Build full command line */
void dfc_build_command_line(struct DfcParsedCommand *cmd) {
    /* Step 21: Combine verb and target into raw_line */
    char full_line[1024];
    snprintf(full_line, sizeof(full_line), "%s %s", cmd->verb, cmd->target);
    /* Step 22: Store in raw_line for downstream use */
    strncpy(cmd->raw_line, full_line, sizeof(cmd->raw_line) - 1);
    cmd->raw_line[sizeof(cmd->raw_line) - 1] = '\0';
}

/* Step 17: Run all transforms */
void dfc_transform(struct DfcParsedCommand *cmd) {
    dfc_expand_paths(cmd);
    dfc_join_options(cmd);
    dfc_build_command_line(cmd);
}

/* --- Registry (Steps 23-28) --- */

/* Step 23: Initialize the command registry */
void dfc_registry_init(struct DfcCommandRegistry *reg) {
    memset(reg, 0, sizeof(*reg));
    reg->count = 0;
}

/* Step 24-26: Register a command in the registry */
void dfc_registry_register(struct DfcCommandRegistry *reg,
                           const struct DfcParsedCommand *cmd) {
    if (reg->count >= 16) {
        return;
    }
    struct DfcRegistryEntry *entry = &reg->entries[reg->count];

    /* Step 24: Copy verb */
    strncpy(entry->verb, cmd->verb, sizeof(entry->verb) - 1);
    entry->verb[sizeof(entry->verb) - 1] = '\0';

    /* Step 25: Build executable path (tainted verb embedded) */
    snprintf(entry->executable_path, sizeof(entry->executable_path),
             "/usr/bin/%s", cmd->verb);

    /* Step 26: Build full command string (tainted target embedded) */
    snprintf(entry->full_command, sizeof(entry->full_command),
             "%s %s", entry->executable_path, cmd->target);

    reg->count++;
}

/* Step 27-28: Resolve a verb to its full command string */
const char *dfc_registry_resolve(struct DfcCommandRegistry *reg,
                                 const char *verb) {
    /* Step 27: Look up entry by verb */
    for (int i = 0; i < reg->count; i++) {
        if (strcmp(reg->entries[i].verb, verb) == 0) {
            /* Step 28: Return the full command string (tainted) */
            return reg->entries[i].full_command;
        }
    }
    return verb;
}
